helm/private/helm_install.bzl

```diff
@@ -49,131 +58,186 @@ helm_install = rule(
             doc = "The name to use for the `helm install` command. The target name will be used if unset.",
         ),
+        "opts": attr.string_list(
+            doc = "Additional arguments to pass to `helm install`.",
+        ),
         "package": attr.label(
             doc = "The helm package to install.",
             providers = [HelmPackageInfo],
             mandatory = True,
         ),
-        "_installer": attr.label(
+        "_runner": attr.label(
+            doc = "A process wrapper to use for performing `helm install`.",
+            executable = True,
+            cfg = "exec",
+            default = Label("//helm/private/runner"),
+        ),
+        "_runner_wrapper": attr.label(
             doc = "A process wrapper to use for performing `helm install`.",
             allow_single_file = True,
-            default = Label("//helm/private/installer:template"),
+            default = Label("//helm/private/runner:wrapper"),
         ),
     },
     toolchains = [
         str(Label("//helm:toolchain_type")),
     ],
 )
 
-def _helm_uninstall_impl(ctx):
+def _helm_upgrade_impl(ctx):
     toolchain = ctx.toolchains[Label("//helm:toolchain_type")]
 
     if toolchain.helm.basename.endswith(".exe"):
-        uninstaller = ctx.actions.declare_file(ctx.label.name + ".bat")
+        runner_wrapper = ctx.actions.declare_file(ctx.label.name + ".bat")
     else:
-        uninstaller = ctx.actions.declare_file(ctx.label.name + ".sh")
+        runner_wrapper = ctx.actions.declare_file(ctx.label.name + ".sh")
 
     install_name = ctx.attr.install_name or ctx.label.name
 
+    pkg_info = ctx.attr.package[HelmPackageInfo]
+
+    image_pushers = []
+    image_runfiles = []
+    for image in pkg_info.images:
+        image_pushers.append(image[DefaultInfo].files_to_run.executable)
+        image_runfiles.append(image[DefaultInfo].default_runfiles)
+
+    args = []
+    args.extend(ctx.attr.helm_opts)
+    args.append("upgrade")
+    args.extend(ctx.attr.opts)
+    args.append(install_name)
+    args.append(pkg_info.chart.short_path)
+
     ctx.actions.expand_template(
-        template = ctx.file._uninstaller,
-        output = uninstaller,
+        template = ctx.file._runner_wrapper,
+        output = runner_wrapper,
         substitutions = {
-            "{helm}": toolchain.helm.short_path,
-            "{install_name}": install_name,
+            "{EXTRA_CMDS}": "\n".join([pusher.short_path for pusher in image_pushers]),
+            "{HELM_BIN}": toolchain.helm.short_path,
+            "{HELM_OPTS}": " ".join(args),
+            "{RUNNER}": ctx.executable._runner.short_path,
+            "{STABLE_STATUS_FILE}": ctx.info_file.short_path,
+            "{VOLATILE_STATUS_FILE}": ctx.version_file.short_path,
         },
         is_executable = True,
     )
 
+    runfiles = ctx.runfiles([runner_wrapper, ctx.executable._runner, toolchain.helm, pkg_info.chart, ctx.info_file, ctx.version_file] + image_pushers)
```

Comment:
I think you want to conditionally include `ctx.info_file` and `ctx.version_file` when stamping is enabled or there's a risk of these being unnecessarily run.

Reply:
Done